### PR TITLE
issue 4411, remove layering violation from cfngen.py/trop.py

### PR DESCRIPTION
**Pull request by lsh-0**
https://github.com/elifesciences/issues/issues/4411

* cfngen, all context wrangling shifted into build_context_* funcs
* isolate each wrangler, lots of assumptions going on in each one
* removed `context.project` access during context creation time
* updated/replaced `context.project` access outside of context creation time with a pared down version of the project data

### CODE REVIEW

**Comment by lsh-0** [comment]
... it actually wasn't that hard, there were just *a lot* of accesses using this pattern. Hopefully the distinction between reading the project data and reading the context data is now clearer.

File: src/project.py [hunk 1]
```diff
@@ -17,16 +17,30 @@ def lst():
...
         'json': core_utils.json_dumps,
         'yaml': core_utils.yaml_dumps,
         # None: core_utils.remove_ordereddict
         None: lambda v: v
     }
+    ensure(output_format in formatters.keys(), "unknown output format %r" % output_format)
     formatter = formatters.get(output_format)
     return formatter(project.project_data(pname))
 
+@task
```
Comment (giorgiosironi) [inline src/project.py]: very useful to introspect the shape of `context`

File: src/masterless.py [hunk 2]
```diff
@@ -114,8 +114,11 @@ def launch(pname, instance_id=None, alt_config='standalone', *repolist):
 @requires_masterless
 def set_versions(stackname, *repolist):
     "call with formula name and a revision, like: builder-private@ab87af78asdf2321431f31"
-    ctx = context_handler.load_context(stackname)
-    repolist = parse_validate_repolist(ctx['project'], *repolist)
+    #ctx = context_handler.load_context(stackname)
+    #repolist = parse_validate_repolist(ctx['project'], *repolist)
+
+    pdata = core.project_data_from_stackname(stackname)
```
Comment (giorgiosironi) [inline src/masterless.py]: uses `elife.yaml` data rather than an existing context for an older instance, but for `masterless` usage this is fine

File: src/masterless.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ def requires_masterless(fn):
     @wraps(fn)
     def wrapper(stackname=None, *args, **kwargs):
         ctx = context_handler.load_context(stackname)
-        ensure(stackname and ctx['project']['aws']['ec2']['masterless'], "this command requires a masterless instance.")
+        ensure(stackname and ctx['aws']['ec2']['masterless'], "this command requires a masterless instance.")
```
Comment (giorgiosironi) [inline src/masterless.py]: shouldn't this be `ctx['ec2']['masterless']`?

File: src/master.py [hunk 4]
```diff
@@ -116,13 +117,13 @@ def remaster(stackname, new_master_stackname):
...
 
-    LOG.info("setting new master address")
-    context = cfngen.set_master_address(context, master_ip) # mutator
-
     LOG.info("upgrading salt client")
     pdata = core.project_data_for_stackname(stackname)
     context['project']['salt'] = pdata['salt']
 
+    LOG.info("setting new master address")
+    cfngen.set_master_address(pdata, context, master_ip) # mutator
```
Comment (giorgiosironi) [inline src/master.py]: not clear in general what it mutates (`context` as opposed to `pdata`), could be the first argument

File: src/master.py [hunk 5]
```diff
@@ -97,15 +97,16 @@ def remaster(stackname, new_master_stackname):
...
-        # TODO: duplicates bad ec2 data wrangling in cfngen.build_context
-        # ec2 == True for some reason, which is completely useless
-        LOG.warn("bad context for stack: %s", stackname)
-        context['ec2'] = {}
-        context['project']['aws']['ec2'] = {}
-
-    if context.get('ec2') == False:
-        LOG.info("ec2 == False, skipping %s", stackname)
+
+    # remove if no longer an issue
```
Comment (giorgiosironi) [inline src/master.py]: no idea on whether this will happen again, good call on commenting it out

Comment (giorgiosironi) [reply src/master.py]: (rather than removing)

File: src/project.py (hunk 1 above)
Comment (giorgiosironi) [reply src/project.py]: should probably support an environment too, e.g. `lax,end2end` to check for stuff that is related only to particular environments like RDS

File: src/buildercore/trop.py [hunk 6]
```diff
@@ -205,18 +200,14 @@ def ec2instance(context, node):
...
     project_ec2 = {
-        "ImageId": lu('project.aws.ec2.ami'),
-        "InstanceType": context['ec2']['type'], # t2.small, m1.medium, etc
+        "ImageId": lu('ec2.ami'),
+        "InstanceType": lu('ec2.type'), # "t2.small", "m1.medium", etc
         "KeyName": Ref(KEYPAIR),
         "SecurityGroupIds": [Ref(SECURITY_GROUP_TITLE)],
-        "SubnetId": subnet_id, # ll: "subnet-1d4eb46a"
+        "SubnetId": subnet_id, # "subnet-1d4eb46a"
         "Tags": instance_tags(context, node),
```
Comment (giorgiosironi) [inline src/buildercore/trop.py]: the `trop` module in general seems to have been simplified

File: src/buildercore/cfngen.py [hunk 7]
```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
...
-    context.update(core.hostname_struct(stackname))
+    # provides: 'domain', 'int_domain', 'subdomain',
+    #           'hostname', 'project_hostname', 'int_project_hostname',
+    #           'full_hostname', 'int_full_hostname'
+    context.update(core.hostname_struct(context['stackname']))
+
+    # project data
+    # preseve some of the project data. all of it is too much
+    keepers = [
+        'salt',
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: later on `salt` should probably be moved to `ec2` as it's only used for that. `formula-*` fields can in theory be used with K8S too so it's the right place.

File: src/buildercore/cfngen.py [hunk 8]
```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
...
+
+    return context
+
+def build_context_aws(pdata, context):
+    if 'aws' not in pdata:
+        return context
+    keepers = [
+        'region',
+        'vpc-id',
+        'subnet-id',
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: maybe the `subnet-*` and `redundant-subnet-*` fields could be moved to `context['ec2']` (not part of this PR)
RDS refers to them only as "allows EC2 instances to connect to me"; not sure if ElastiCache is the same.

Comment (giorgiosironi) [reply src/buildercore/cfngen.py]: `context['elasticache']` has its own `subnets` field so yes, we'll move them.
`region` and `vpc-id` are so general for AWS that they are already in the right place.

Comment (giorgiosironi) [reply src/buildercore/cfngen.py]: `context['elb']` will need to get them from `aws.elb.subnets`, as it doesn't at the moment

File: src/buildercore/cfngen.py [hunk 9]
```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
...
+        build_context_vault,
+    ]
+
+    # exceptions to the rule ...
+    wrangler_list = [project_wrangler] + wrangler_list
+
+    for wrangler in wrangler_list:
+        # so functions can't modify the context in-place or
+        # reference a bit of project_data and then change it
+        context = wrangler(deepcopy(project_data), deepcopy(context))
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: that's good, true immutability. Ideally they wouldn't need `context` as an input, right? But it would be difficult to merge them all.

Comment (giorgiosironi) [reply src/buildercore/cfngen.py]: `project_name` and `stackname` are the only fields that are read from the context in this module, so I'd try to simplify that to remove `deepcopy(context)` as an argument. Would be a follow-up when this has been integrated.

File: src/buildercore/bootstrap.py [hunk 10]
```diff
@@ -460,23 +457,23 @@ def update_ec2_stack(stackname, context, concurrency=None, formula_revisions=Non
     script that can be downloaded from the web and then very conveniently
     installs it's own dependencies. Once Salt is installed we give it an ID
     (the given `stackname`), the address of the master server """
+
+    # no longer a snapshot of the project data at creation time, now it's
+    # mostly just top-level project attributes as we need them.
     pdata = context['project']
```
Comment (giorgiosironi) [inline src/buildercore/bootstrap.py]: should be renamed from `pdata` to something like `project_repositories` to avoid confusion with the "snapshot of the project data"

File: src/master.py (hunk 4 above)
Comment (lsh-0) [reply src/master.py]: I'll elaborate

File: src/buildercore/trop.py (hunk 6 above)
Comment (lsh-0) [reply src/buildercore/trop.py]: a little nip and tuck

File: src/buildercore/cfngen.py [hunk 11]
```diff
@@ -222,125 +294,130 @@ def build_context_rds(context, existing_context):
...
 
-def build_context_gcp(context, parameterize):
-    if 'gcs' in context['project']['aws']:
+def build_context_gcp(pdata, context):
+    context['gcs'] = False
+    if 'gcs' in pdata['aws']:
         context['gcs'] = OrderedDict()
-        for bucket_template_name, options in context['project']['aws']['gcs'].items():
-            bucket_name = parameterize(bucket_template_name)
+        for bucket_template_name, options in pdata['aws']['gcs'].items():
```
[the remark this answers is not in the document]
Comment (lsh-0) [reply src/buildercore/cfngen.py]: yes, we have another ticket (I think?) to move `gcs` out from under `aws`.

File: src/buildercore/cfngen.py (hunk 7 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: this snippet was copied from trop.py and trop.py pares down even further

File: src/buildercore/cfngen.py (hunk 8 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: :+1:

File: src/buildercore/cfngen.py (hunk 9 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: I can see pros and cons, but even though I commented it out, I think shuffling the order of the wranglers is a good idea to prevent ordering dependencies. Removing the context as a parameter would obviate the shuffling.

**Comment by lsh-0** [comment]
@giorgiosironi , I think I addressed your questions. if you're happy please merge.

File: src/masterless.py (hunk 3 above)
Comment (lsh-0) [reply src/masterless.py]: hm, missed this one. you might be right.

Comment (lsh-0) [reply src/masterless.py]: yes, it happens here: https://github.com/elifesciences/builder/blob/3886100f47ecf87c0c39019c1b001cb345bb92c6/src/buildercore/cfngen.py#L265

I have it all checked out, I'll fix it up

**Comment by giorgiosironi** [comment]
I am finding some problems in the `masterless` module, will do some full testing on it.

**Comment by lsh-0** [comment]
no worries - let me know if I can help

**Comment by giorgiosironi** [comment]
We need to expand `integration_tests` to cover also some basic `masterless` tasks; I am afraid of the test suite slowdown however.

**Comment by lsh-0** [comment]
well done, I wonder how that typo made it past testing

**Comment by lsh-0** [comment]
can we run those integration tests in parallel?

**Comment by giorgiosironi** [comment]
> can we run those integration tests in parallel?

Probably, but there is some [outstanding problem to solve I'll get on today](https://github.com/elifesciences/builder/pull/290)
